lib/generators.nix

```diff
@@ -402,6 +402,139 @@ rec {
 ${expr "" v}
 </plist>'';
 
+  /* Generates a libconfig configuration file.
+   *
+   * Currently quite close to fully-featured, though there are probably some
+   * oversights not covered by the test (./tests/misc.nix) and automated validator. (//pkgs/pkgs-lib)
+   *
+   * https://hyperrealm.github.io/libconfig/libconfig_manual.html
+   */
+  toLibconfig = { }:
+    v:
+    let
+      isFloat = builtins.isFloat or (x: false);
+      isHexString = str: builtins.match "^[+-]?0x[0-9a-fA-F]+$" str != null;
+      isOctalString = str: builtins.match "^[+-]?0[1-7][0-7]*$" str != null;
+      isLongString = str: builtins.match "^[+-]?[0-9]+L$" str != null;
+      isNumber = x: (isHexString x) || (isOctalString x) || (isLongString x);
+
+      # > An array may have zero or more elements, but the elements must
+      # > all be *scalar* values of the same type.
+      isArray = x: let
+        typed = builtins.groupBy type x;
+        filtered = builtins.removeAttrs typed [ "null" "aggregate" ];
+      in
+        # > ... *scalar* values..
+        !(typed ? aggregate) &&
+        # > ..of the same type
+        lib.length (builtins.attrNames filtered) <= 1;
+
+      type = x:
+        with builtins;
+        if x == null then
+          "null"
+        # > A scalar type (integer, 64-bit integer, floating point, boolean, or string)
+        else if isBool x then
+          "bool"
+        else if isInt x then
+          "int"
+        else if isFloat x then
+          "float"
+        else if isString x then
+          if isHexString x then
+            "hex"
+          else if isOctalString x then
+            "octal"
+          else if isLongString x then
+            "long"
+          else
+            "string"
+        else
+          # > An aggregate type (a group, array, or list),
+          "aggregate";
+
+      expr = ind: x:
+        with builtins;
+        if x == null then
+          ""
+        else if isBool x then
+          bool ind x
+        else if isInt x then
+          num ind x
+        else if isFloat x then
+          num ind x
+        else if isString x then
+          if isNumber x then
```

Comment:
Please don't assign meaning without context. Such assumptions are likely to lead to bugs.

```
"12A" -> "12A"
"12B" -> "12B"
...
"12L" -> 12
```

This is not how Nix works. Strings are strings. If you want a number, you write a number, or call a function that returns one.

Instead, you could represent these special with special wrappers instead `mkOctal`, etc, but do you really need multiple ways to write the same number?

Reply:
> but do you really need multiple ways to write the same number?

Spec says there are a few scalar types and @anund previously said:
> [~](https://github.com/NixOS/nixpkgs/pull/167388#discussion_r933727593) This [couldn't] currently generate valid config for the example. :(

Other then that `mk{Octal|…}` seems reasonable but annoying, _future note:_ make it return just a simple `{ _type = …; … }` attrset.

I'll get to this Eventually.

Reply:
Playing with https://github.com/PixlOne/logiops/blob/master/logid.example.cfg as an example.

```
devices = [
  {
    #<snip> out direct translations
    buttons = [
      {
        cid: {toLibConfig.mkOctal: "0xc3"};
        action =
          {
            type: "Gestures";
            gestures: [
              {
                direction: "Up";
                mode: "OnRelease";
                action =
                {
                  type: "Keypress";
                  keys: {toLibConfig.mkArray: ["KEY_UP"]};
                };
              }
            ];
          };
      }
    ];
  }
];
```
Is that the sort of wrappers you have in mind @roberth?

Reply:
If https://github.com/NixOS/nix/issues/7578 is implemented, the hex and octal numbers won't need any sort of escaping or wrapping. That is, unless the program parsing the config makes decisions based on the syntax used, which would be decimal regardless of what base the module user used to configured the number.

I don't think we have many examples of wrappers yet, but in the module system we generally use a representation that's an attrset with a `_type` attribute. For example:

```nix
  mkOverride = priority: content:
    { _type = "override";
      inherit priority content;
    };
```

I don't think we have any examples of formats using this approach. (I kind of expected at least something RFC42-like to use this, so sorry for my brevity earlier)
We could add a general `mkUnescape` to `lib`, but otherwise I don't know exactly where to put it. Perhaps we could have `types.foo.type` and also `types.foo.lib`?
The meaning of this `mkUnescape` is that it would take a string and concatenate it directly into the config file without any processing (except perhaps re-indenting, for some formats). This is in contrast to regular strings, which should be escaped to become strings in the target config language. This means that the string is reinterpreted according to the target configuration language syntax, and this would allow a user to specify any syntax, assuming it's going to be valid in the end. This includes such things as hex: `lib.mkUnescape "0xc3"` would do the job.

cc @infinisil

Reply:
That's not to say that a `formats.libconfig.mkOctal` wouldn't be nice to have.

Reply:
A quick survey: octal is 0o[0-7]+ in python/haskell, 0[0-7]+ in java/c/golang/libconfig. This might be a format that requires a wrapper. (or a pattern of read as octal write as decimal?)